.github/workflows/main.yml

```diff
@@ -11,16 +11,69 @@ on:
 
 jobs:
   test:
-    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        libvirt: ['3.1.0', '5.1.0']
+        include:
+          - libvirt: '3.1.0'
+            os: 'ubuntu-18.04'
+          - libvirt: '5.1.0'
+            os: 'ubuntu-18.04'
+
+    runs-on: ${{ matrix.os }}
+    env:
+      CCACHE_TEMPDIR: /tmp/.ccache-temp
+      EXT: xz
+      LIBVIRT: ${{ matrix.libvirt }}
 
     steps:
       - uses: actions/checkout@v2 # respository path: $GITHUB_WORKSPACE
       - uses: actions/setup-go@v2
         with:
           go-version: 1.16.2
 
-      - name: Install dependencies
-        run: go get -u golang.org/x/lint/golint
+      - name: Install golint and goyacc
+        run: |
+          go get -u golang.org/x/lint/golint
+          go get golang.org/x/tools/cmd/goyacc
+
+      - name: Enable apt source package downloads and update index
+        run: |
+          sudo sed -i '/deb-src/s/^# //' /etc/apt/sources.list
+          sudo apt update
+
+      - name: Build libvirt from source and install it
+        run: |
+          sudo apt-get -qqy build-dep libvirt
+          sudo apt-get -qqy install curl ccache qemu-system-x86
+          sudo mkdir -p /usr/src && sudo chown $(id -u) /usr/src
+          curl -O -s https://libvirt.org/sources/libvirt-${LIBVIRT}.tar.${EXT}
+          tar -C /usr/src -xf libvirt-${LIBVIRT}.tar.${EXT}
+          pushd /usr/src/libvirt-${LIBVIRT}
+          ccache --show-stats
+          env PATH=/usr/lib/ccache:$PATH \
+             ./configure --prefix=/usr --localstatedir=/var --sysconfdir=/etc \
+                         --without-polkit \
+                         --without-esx --without-vbox --without-xen --without-libxl --without-lxc \
+                         --with-qemu
+          make
+          sudo make install
+          ccache --show-stats
+          popd
+          sudo libvirtd -d -l -f $GITHUB_WORKSPACE/.github/testdata/libvirtd.conf
+          sudo virtlogd -d || true
+
+      - name: Setup test artifacts
+        env:
+          TESTDATA: ${{ github.workspace }}/.github/testdata
+        run: |
+          go get -d ./...
+          sudo qemu-img create -f raw -o size=10M /var/lib/libvirt/images/test.raw
```

Comment:
Why not put this in `Makefile`, so its easy for devs to find it and run tests locally before GH actions run: `sudo make test` or `sudo make ci-test`?

Reply:
From what I've read (i.e. a very quick google), it's not possible to run these tests locally without installing another tool that helps allow you to do so.  If someone wants to check their branch passes before opening the PR, they must push their branch, and then use the github GUI to manually trigger a run.  I can add something to the README to that effect.

Reply:
It would be a nice enhancement to have things organized in a way where you can manually run the tests using a container locally; however, I don't have the bandwidth at the moment to work on something like that and would rather spend that time trying to get these things updated to be able to test more modern versions of libvirt.  I haven't found the workflow of pushing the branch and clicking the button all the burdensome or particularly slow (probably because the tests themselves take a bit simply because they compiles libvirt from source).

Reply:
^--- I lack write access, so can't resolve this thread :-).